### DISCUSSION TITLE
Lua filter to extract document metadata and include in the body

### DISCUSSION THREAD

**Discussion by ratnanil** [quarto-dev/quarto-cli#4042] [Q&A] [answered]
I'm trying to write a Lua filter which grabs data from the document's YAML header and injects it into the document's body. Specifically, I want to add a footnote to all level 1 headers. My index.qmd looks as follows:

```
---
title: "MyTitle"
mymeta:
  key1: val1
  key2: val2
filters:
  - add_to_title.lua
---

# A level 1 header
``` 

So far, I've written the following Lua filter (`add_to_title.lua`), which gets me far, but not quite there yet. Note that I would want to replace `"placeholder"` with data from the YAML header. 

```{lua}
function Header(element)
    if element.level == 1 then

        element.content:extend({pandoc.Note("placeholder")})
        return pandoc.Header(1, element.content)

    else
        return nil
    end
end
``` 

Which renders to the following. Ideally, `placeholder` would be replaced with `key1: value1, key2: value2`:

![image](https://user-images.githubusercontent.com/12532091/213831389-a5c9b69b-a743-484e-a230-a6df5e553d46.png)

I've tried all sorts of ways to access the document's YAML header / metadata, with no success. How do I achieve this?

**Comment by cscheid** [comment, marked as answer]
Try this:

```
local my_key_values = {}

function Meta(meta)
  for key, value in pairs(meta.mymeta) do
    my_key_values[key] = value
  end
end

-- thank you copilot
local function by_sorted_keys(t)
  local keys = {}
  for k in pairs(t) do table.insert(keys, k) end
  table.sort(keys)
  local i = 0      -- iterator variable
  local iter = function ()   -- iterator function
    i = i + 1
    if keys[i] == nil then return nil
    else return keys[i], t[keys[i]]
    end
  end
  return iter
end

function Header(element)
  if element.level == 1 then
    local plain = pandoc.Para({})

    for key, value in by_sorted_keys(my_key_values) do
      plain.content:insert(pandoc.Strong(pandoc.Str(key .. ": ")))
      plain.content:extend(value)
      plain.content:insert(pandoc.SoftBreak())
    end
    local placeholder = pandoc.Note(plain)
    element.content:insert(placeholder)

    return element
  else
    return nil
  end
end

return {
  {Meta = Meta},
  {Header = Header}
}
```

The notable trick here is that you need to create a filter that _first_ finds the meta information, and _then_ uses it.

The result I get from this filter is

<img width="363" alt="image" src="https://user-images.githubusercontent.com/285675/213872485-a93aed9a-f99e-4090-b193-e23d70adab1c.png">

**Reply by pagiraud** [reply]
I just don’t get how that works.

I have the same .qmd file than @ratnanil and my filter content is only:

```
local my_key_values = {}

function Meta(meta)
  for key, value in pairs(meta.mymeta) do
    my_key_values[key] = value
  end
end

print(my_key_values[key1])

```

And the print says `nil`… So the variable is not populated. How can I populate the variable?

Thanks in advance.

EDIT: I understand this is linked to variable scope, but I’d like to know more… Meta and Header are not the same function…

**Reply by cscheid** [reply]
It's not about variable scope. It's that you're issuing the print() statement _before_ the filter runs. Try putting a print statement inside the `Meta` function to see what's happening.

**Reply by pagiraud** [reply]
Oh thank you so much for your reply. I had already seen that the value was correct inside the `Meta` function (actually, I had seen it before stumbling upon this thread, trying to solve my problem) but assumed it was a scope problem. I would have never thought by myself about your explanation, but it makes sense now !

**Comment by ratnanil** [comment]
It works perfectly. I didn't realize how far I was from the solution... sorry for this! But thank you so much for solving my problem none the less!
